### ISSUE TITLE
Automatically ensure validity invariants of unsafe operations

### ISSUE THREAD

**Issue by celinval** [model-checking/kani#2998]
Requested feature: Kani should automatically check if unsafe operations can break validity invariants. This follows the concept of validity from https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html.
Use case: Checking that unsafe code doesn't trigger UB.
Link to relevant documentation (Rust reference, Nomicon, RFC):

Test case:
```rust
#[kani::requires(!ptr.is_null())]
#[kani::ensures(!result.as_ptr().is_null())] // Compiler warns that this should always be nonnull().
pub unsafe fn new_unchecked<T>(ptr: *mut T) -> NonNull<T> {
    NonNull::<T>::new_unchecked(ptr)
}
```

The ensure clause here might be redundant, since the compiler might decide to prune it, i.e., you cannot write Rust code to check for UB occurrence.

**Comment by celinval** [comment]
Note for self: We should use implement https://github.com/rust-lang/project-stable-mir/issues/58 and use the valid range from Scalar type. We would need to traverse structures to add proper checks. Maybe experiment with transforming StableMIR instead of regular MIR. This would play well since we want the monomorphic version to add validity checks. Here are a few good places to see how rustc / MIRI handle these things:
- ValidityVisitor: https://github.com/rust-lang/rust/blob/master/compiler/rustc_const_eval/src/interpret/validity.rs#L655
- List of errors: https://github.com/rust-lang/rust/blob/d8b00690ec3aa2d2fb0d3eb00e862195501186f9/compiler/rustc_const_eval/messages.ftl#L416-L474 (Some are not redundant for this issue)
